ocaml/xapi/xapi_sdn_controller.ml

```diff
@@ -0,0 +1,54 @@
+(*
+ * Copyright (C) 2006-2016 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Stdext
+open Pervasiveext
+open Xstringext
+open Helpers
+open Listext
+open Client
+open Stdext.Threadext
+open Unixext
+open Network
+
+module D = Debug.Make(struct let name="xapi" end)
+open D
+
+let protocol_to_string = function
+  | `ssl -> "ssl"
+  | `pssl -> "pssl"
+  | _ -> "ssl"
+
+let introduce ~__context ~protocol ~address ~port =
+  let dbg = Context.string_of_task __context in
+  match Net.Bridge.get_kind dbg () with
+  | Network_interface.Openvswitch ->
+    if List.length (Db.SDN_controller.get_all ~__context) > 0 then
+      raise (Api_errors.Server_error(Api_errors.operation_not_allowed, ["SDN controller has been configured. Please forget it first."]));
+      if address <> "" then
+        Helpers.assert_is_valid_ip `ipv4 "address" address;
+      if port <> 0L then
+        Helpers.assert_is_valid_tcp_udp_port (Int64.to_int port) "port";
+      let r = Ref.make () and uuid = Uuid.make_uuid () in
+      Db.SDN_controller.create ~__context ~ref:r ~uuid:(Uuid.to_string uuid) ~protocol ~address ~port;
+      List.iter (fun host -> Helpers.update_vswitch_controller ~__context ~host) (Db.Host.get_all ~__context);r
+  | _ -> raise (Api_errors.Server_error(Api_errors.operation_not_allowed, ["host not configured for vswitch operation"]))
+
+let forget ~__context ~self =
+  let dbg = Context.string_of_task __context in
+  match Net.Bridge.get_kind dbg () with
+  | Network_interface.Openvswitch ->
+      Db.SDN_controller.destroy ~__context ~self;
+      List.iter (fun host -> Helpers.update_vswitch_controller ~__context ~host) (Db.Host.get_all ~__context);
```

Comment:
...but we should only call `Helpers.update_vswitch_controller` if the backend is OVS. In other words, the match becomes an if-statement here, and we won't raise an exception.